AzureMonitor: Migrate to backend checkHealth API

aangelisc opened the pull request:
**What this PR does / why we need it**:

Migrates the Azure Monitor datasource to make use of the backend CheckHealth API rather than implementing health checks in the frontend. Also adds a health check for Azure Resource Graph which did not previously exist.

- Azure Monitor metrics health check is carried out by ensuring the monitor endpoint is reachable without errors.
- Azure Log Analytics health check is carried out by ensuring the metadata for the default workspace is reachable without errors.
- Azure Resource Graph health check is carried out by ensuring that a simple query can be executed without errors.

All frontend health checks are removed. However, as the main datasource class does not extend `DataSourceWithBackend`, the `testDatasource` function is called via the Azure Monitor metrics pseudo-datasource. 

**Which issue(s) this PR fixes**:

Fixes #50236 

**Special notes for your reviewer**:

Health check before changes:
![Screenshot_2022-06-15 11 07 32_6JaVz0](https://user-images.githubusercontent.com/15019026/173803455-22dcd10e-dd29-4fa3-8414-110056f24d13.png)

Health check after changes:
![Screenshot_2022-06-15 11 09 49_nhsWaz](https://user-images.githubusercontent.com/15019026/173803497-7918c8a0-154a-4750-9c0a-d4aff67f6dac.png)

Health check after changes with error:
![Screenshot_2022-06-15 11 10 46_efzVQY](https://user-images.githubusercontent.com/15019026/173803549-bfb6ee50-c81b-423e-8fce-c18018632be3.png)

aangelisc committed:
Add check health functions for each datasource and generic checkHealth function

aangelisc committed:
Log backend errors

aangelisc committed:
Update testDatasource function

- Remove unused testDatasource functions from pseudo datasources

sunker reviewed (commented):
Great stuff! 

As you point out, we could get rid of some boilerplate code if the datasource file extends DataSourceWithBackend instead of DataSourceApi. For that reason, I think it's worth investigating how big of an effort that refactoring would be before merging this. It's a refactoring that we should do at some point anyway, so we might as well start now. 

I think it would be good if the backend health check had the same behavior as the frontend check has. I.e it should always test all three endpoints before returning a response. The message returned to the frontend should indicate which out of the three APIs it was that failed.

aangelisc commented:
> As you point out, we could get rid of some boilerplate code if the datasource file extends DataSourceWithBackend instead of DataSourceApi. For that reason, I think it's worth investigating how big of an effort that refactoring would be before merging this. It's a refactoring that we should do at some point anyway, so we might as well start now.

This is a good shout - I'll investigate this today.

> I think it would be good if the backend health check had the same behavior as the frontend check has. I.e it should always test all three endpoints before returning a response. The message returned to the frontend should indicate which out of the three APIs it was that failed.

A good point. I wonder if we should return the full error to the frontend or leave it as a backend log - what are your thoughts on this?

sunker commented:
> A good point. I wonder if we should return the full error to the frontend or leave it as a backend log - what are your thoughts on this?

My preference would be to log the full error in the backend and just pass the test status for each API to the frontend.

aangelisc committed:
Switch datasource to extend DataSourceWithBackend

aangelisc committed:
Improve errors and responses from health endpoint

aangelisc committed:
Fix backend lint issues

aangelisc committed:
Remove unneeded frontend tests

aangelisc committed:
Remove unused/unnecessary datasource methods

aangelisc committed:
Update types

aangelisc committed:
Improve message construction

aangelisc committed:
Stubbing out checkHealth tests

aangelisc committed:
Update tests

- Remove comments
- Simplify structure

aangelisc committed:
Update log analytics health check to query data rather than retrieve workspace metadata

aangelisc committed:
Fix lint issue

aangelisc commented on pkg/tsdb/azuremonitor/azuremonitor.go:
```diff
@@ -161,3 +164,179 @@ func (s *Service) newQueryMux() *datasource.QueryTypeMux {
+		}
+
+		if len(target.Value) == 0 {
+			return nil, errors.New("no default workspace found")
```
I personally don't think this case should be an error but rather a warning to the user that there is no available log analytics workspace. I'd like thoughts on this as currently the `BackendCheckHealthResult` status only supports `UNKNOWN`, `ERROR`, `OK`. I think an additional warning status should be included allowing for a warning alert to be displayed in the frontend rather than an error.

aangelisc replied:
I will test this and get back to you 😊

aangelisc commented on pkg/tsdb/azuremonitor/azuremonitor.go:
```diff
@@ -161,3 +164,179 @@ func (s *Service) newQueryMux() *datasource.QueryTypeMux {
+	}
+
+	body, err := json.Marshal(map[string]interface{}{
+		"query": "AzureActivity | limit 1",
```
I'm not 100% certain on if this will always be present in every Workspace - may need to reach out to MSFT for confirmation.

aangelisc committed:
Fix frontend lint issues

aangelisc committed:
Merge branch 'main' into 50236-azure-monitor-check-health-migration

aangelisc committed:
Merge branch 'main' into 50236-azure-monitor-check-health-migration

andresmgot reviewed (commented):
Good work here! I have some suggestions

aangelisc commented on public/app/plugins/datasource/grafana-azure-monitor-datasource/datasource.ts:
```diff
@@ -142,48 +146,58 @@ export default class Datasource extends DataSourceApi<AzureMonitorQuery, AzureDa
-    return await Promise.all(promises).then((results) => {
-      let status: 'success' | 'error' = 'success';
-      let message = '';
+    if (authType === 'clientsecret') {
+      if (!this.isValidConfigField(this.instanceSettings.jsonData.tenantId)) {
+        return {
+          status: 'error',
+          message: 'The Tenant Id field is required.',
+        };
+      }
```
I kept this from the original `testDatasource` methods. Setting the field to `required` seems to have no effect. Unsure if that's because they're using legacy forms (should probably be updated as well). The error handling could be improved to provide the appropriate information though I think the form should be updated to use the newer components where the `required` prop should have an effect..

andresmgot replied:
> though I think the form should be updated to use the newer components where the `required` prop should have an effect

I agree but in the meantime, I would remove this code. It's not worth it to keep a reference to `instanceSettings` and everything just to validate a mandatory input (which is obvious anyway).

aangelisc committed:
Update pkg/tsdb/azuremonitor/azuremonitor.go

Co-authored-by: Andres Martinez Gotor <andres.martinez@grafana.com>

aangelisc committed:
Updates based on PR comments

- Don't use deprecated default workspace field
- Handle situation if no workspace is found by notifying user
- Correctly handle health responses

aangelisc committed:
Merge branch '50236-azure-monitor-check-health-migration' of github.com:grafana/grafana into 50236-azure-monitor-check-health-migration

aangelisc committed:
Remove debug line

aangelisc commented on pkg/tsdb/azuremonitor/azuremonitor.go:
```diff
@@ -161,3 +164,179 @@ func (s *Service) newQueryMux() *datasource.QueryTypeMux {
+		return nil, err
+	}
+
+	res, err := dsInfo.Services["Azure Monitor"].HTTPClient.Do(request)
```
I'll look at switching to this instead..

andresmgot replied:
is this still in progress? if you see that it's not worth it, we can at least export the API version that we are using the the data source and use it here.

aangelisc replied:
Apologies, forgot to comment on this. I opted to stick with the original method as on investigation it seemed simpler and easier to test. Exporting the API version is a good shout. I've added this in.

andresmgot commented on pkg/tsdb/azuremonitor/azuremonitor.go:
```diff
@@ -161,3 +164,179 @@ func (s *Service) newQueryMux() *datasource.QueryTypeMux {
+			if err != nil {
+				return nil, err
+			}
+			backend.Logger.Error(string(body))
```
> * In case anything sensitive is exposed

that was my point above, we are just using the credentials that the user is giving us to create a request that they could create themselves. If there is sensitive information, they have access to that information anyway, there is no point on hiding it.

> The Azure errors can be quite verbose and I don't think this will always present very well in the UI. Happy to have a discussion about this.

I'd rather have a verbose error than no error at all so you don't know what's happening. If the user wants to fix the error, they need to access the Grafana logs and looks to the verbose error anyway, which is even more difficult.

If you still want to discuss this we can jump into a call!

andresmgot commented on pkg/tsdb/azuremonitor/azuremonitor.go:
```diff
@@ -161,3 +164,185 @@ func (s *Service) newQueryMux() *datasource.QueryTypeMux {
+
+	return &backend.CheckHealthResult{
+		Status:  status,
+		Message: fmt.Sprintf("1. %s\n2. %s\n3. %s", metricsLog, logAnalyticsLog, graphLog),
```
btw, are the `\n` being interpreted?

aangelisc replied:
They are not interpreted in the frontend

andresmgot replied:
Anyway, thinking about this, I think that the message is too repetitive (1. Successfully connected ... 2. Successfully connected... etc). We can probably simplify it. My suggestion is

- If the three checks pass, simply say "Successfully connected to Azure Monitor"
- If some check fails then we list each element. I noticed that you can specify "Details" when returning a message. This is helpful for this use case (and properly interprets `\n`). I tried:

```
	return &backend.CheckHealthResult{
		Status:  backend.HealthStatusError,
		Message: "Some health check failed",
		JSONDetails: []byte(
			`{"verboseMessage":"1. Successfuly connected to Azure Monitor Metrics\n2. Error connecting to Azure Resource Graph endpoint: Boom!"}`,
		),
	}, nil
```

and got:

![Screenshot from 2022-06-15 16-59-41](https://user-images.githubusercontent.com/4025665/173859864-0124214c-67ae-49b8-94bc-b235a0df678c.png)

andresmgot commented:
can you update the original comment? and also add a screenshot to see the final result?

aangelisc committed:
Make use of defined api versions

aangelisc committed:
Remove field validation functions

aangelisc committed:
Expose errors in frontend

aangelisc committed:
Merge branch 'main' into 50236-azure-monitor-check-health-migration

# Conflicts:
#	public/app/plugins/datasource/grafana-azure-monitor-datasource/azure_monitor/azure_monitor_datasource.ts

aangelisc committed:
Update errors and tests

andresmgot reviewed (commented):
Thanks for the changes! Some more minor suggestions :)

aangelisc committed:
Remove instanceSettings

aangelisc committed:
Update error handling

andresmgot commented on pkg/tsdb/azuremonitor/azuremonitor.go:
```diff
@@ -161,3 +164,192 @@ func (s *Service) newQueryMux() *datasource.QueryTypeMux {
+		Status:  status,
+		Message: "One or more health checks failed. See details below.",
+		JSONDetails: []byte(
+			fmt.Sprintf(`{"verboseMessage": "1. %s\n2. %s\n3.%s" }`, metricsLog, logAnalyticsLog, graphLog),
```
```suggestion
			fmt.Sprintf(`{"verboseMessage": "1. %s\n2. %s\n3. %s" }`, metricsLog, logAnalyticsLog, graphLog),
```
:)

andresmgot commented on pkg/tsdb/azuremonitor/azuremonitor.go:
```diff
@@ -161,3 +164,185 @@ func (s *Service) newQueryMux() *datasource.QueryTypeMux {
+	metricsRes, err := checkAzureMonitorMetricsHealth(dsInfo)
+	if err != nil || metricsRes.StatusCode != 200 {
+		status = backend.HealthStatusError
+		if err != nil {
```
you are still working on this, right?

aangelisc committed:
Improve error handling and presentation

aangelisc committed:
Update tests and correctly check error type

andresmgot commented on pkg/tsdb/azuremonitor/azuremonitor.go:
```diff
@@ -277,7 +278,11 @@ func (s *Service) CheckHealth(ctx context.Context, req *backend.CheckHealthReque
 		status = backend.HealthStatusError
 		if err != nil {
-			return nil, err
+			if _, ok := err.(types.AzureError); ok {
```
the idiomatic way of doing this is with the method `.Is` (e.g. `errors.Is(err, types.AzureError)`. For that, you need to change how you return the error and how you define it (see the other comments).

andresmgot commented on pkg/tsdb/azuremonitor/types/types.go:
```diff
@@ -213,3 +213,7 @@ type LogAnalyticsWorkspaceResponse struct {
 	RetentionInDays                 int                             `json:"retentionInDays"`
 }
+
+type AzureError interface {
+	Error() string
+}
```
You can define this as an error (I would also name the error something more specific):

```
var AzureHealthCheckError = errors.New("health check failed")
```

You can see this approach many times in our [code base](https://github.com/grafana/grafana/blob/main/pkg/models/org.go#L10)

andresmgot commented on pkg/tsdb/azuremonitor/azuremonitor.go:
```diff
@@ -188,7 +189,7 @@ func checkAzureMonitorMetricsHealth(dsInfo types.DatasourceInfo) (*http.Response
 	res, err := dsInfo.Services["Azure Monitor"].HTTPClient.Do(request)
 	if err != nil {
-		return nil, err
+		return nil, types.AzureError(err)
```
if you define the error as an AzureError, you can return wrap it with `%w` so you can check its type later:

```go
			return nil, fmt.Errorf("%w: %v", types.AzureError, err)
```

andresmgot reviewed (approved):
Great improvement here! Thank you for your patience during the review :) I have one last small comment regarding the error definition but apart from that we are good to go.

aangelisc committed:
Refactor AzureHealthCheckError and update tests

aangelisc committed:
Merge branch 'main' into 50236-azure-monitor-check-health-migration

aangelisc committed:
Fix lint errors
